vpr/src/route/check_route.cpp

```diff
@@ -394,7 +396,7 @@ static bool check_adjacent(int from_node, int to_node, bool is_flat) {
             break;
 
         case OPIN:
-            from_grid_type = device_ctx.grid.get_physical_type(from_xlow, from_ylow);
+            from_grid_type = device_ctx.grid.get_physical_type({from_xlow, from_ylow, from_layer});
```

Comment:
1. check_adjacent should check that the layer numbers are no more than one apart for connected elements (add this check).
2. There should also be checks for all the rr_nodes used in the routing having layer numbers within the grid (should be right by checks that x and y are within the grid.  Check if we have these. (probably in a function called check_node or some such).